pyshtools/shio/icgem.py

```diff
@@ -0,0 +1,188 @@
+    harmonic coefficients.
+
+    Returns
+    -------
+    cilm : array
+        Array with the coefficients with the shape (2, lmax + 1, lmax + 1)
+        for the given epoch.
+    gm : float
+        Standard gravitational constant of the model, in m**3/s**2.
+    r0 : float
+        Reference radius of the model, in meters.
+    errors : array, optional
+        Array with the errors of the coefficients with the shape
+        (2, lmax + 1, lmax + 1) for the given epoch.
+
+    Parameters
+    ----------
+    filename : str
+        The ascii-formatted filename containing the spherical harmonic coefficients.
+    errors : str, optional
+        Which errors to read. Can be either "calibrated", "formal" or
+        None. Default is None.
+    lmax : int, optional
+        Maximum degree to read from the file. If lmax is None, less than 0, or
+        greater than lmax_model, the maximum degree of the model will be used.
+    epoch : str or float, optional
+        The epoch time to calculate time-variable coefficients in YYYYMMDD.DD
+        format. If None then reference epoch t0 of the model will be used.
+        If format of the file is 'icgem2.0' then epoch must be specified.
+    """
+
+    # read header
+    header = {}
+    header_keys = ['modelname', 'product_type', 'earth_gravity_constant',
+                   'gravity_constant', 'radius', 'max_degree', 'errors',
+                   'tide_system', 'norm', 'format']
+
+    with open(filename, 'r') as f:
+        for line in f:
+            if 'end_of_head' in line:
+                break
+            for key in header_keys:
+                if key in line:
+                    header[key] = line.strip().split()[1]
+
+        if header['product_type'] != 'gravity_field':
+            raise ValueError('This function reads only gravity_field data product.')
+
+        is_v2 = False
+        if 'format' in header and header['format'] == 'icgem2.0':
+            is_v2 = True
+
+        if epoch is None and is_v2:
+            raise ValueError('Epoch must be specified for the "icgem2.0" format.')
+        elif epoch is not None:
+            epoch = _yyyymmdd_to_year_fraction(epoch)
+
+        if 'earth_gravity_constant' in header:
+            gravity_constant = float(header['earth_gravity_constant'])
+        elif 'gravity_constant' in header:
+            gravity_constant = float(header['gravity_constant'])
+        else:
+            raise ValueError('No standard gravitational constant in the header')
+
+        radius = float(header['radius'])
+
+        lmax_model = int(header['max_degree'])
+        if lmax is None or lmax < 0 or lmax > lmax_model:
+            lmax = lmax_model
+
+        if errors is not None:
+            valid_err = ('calibrated', 'formal', 'calibrated_and_formal')
+            if header['errors'] == 'no':
+                raise ValueError('This model has no errors.')
+            elif errors not in valid_err[:-1]:
```

Comment:
No, 'calibrated_and_formal' is not implemented.

'calibrated', 'formal' and 'calibrated_and_formal' are parameters in the ICGEM formatted file for keyword 'errors' which has nothing to do with the parameter `errors` of the function.

'calibrated' or 'formal' mean there are only two columns with the errors - 5 and 6, which will be read if  `errors="calibrated"` or `errors="formal"` respectively. 

'calibrated_and_formal', in the other hand, means there are four columns in the file: 5, 6, 7 and 8. 5 and 6 stand for calibrated errors and 7 and 8 stand for formal. So, the function will return calibrated errors if `errors="calibrated"` or formal errors if `errors="formal"` as before.

I don't know if returning of both calibrated and formal errors is really needed. If it does, the I will write it.


This part is checking that `errors` has valid values and the given value to this parameter is actually exists in the file.